lib/matplotlib/_mathtext.py

```diff
@@ -66,6 +66,83 @@ def get_unicode_index(symbol, math=False):  # Publicly exported.
+VectorParse.__doc__ = r"""
+The namedtuple type returned by ``MathTextParser("path").parse(...)``.
+
+This tuple contains the global metrics (*width*, *height*, *depth*), a list of
+*glyphs* (including their positions) and of *rect*\angles.
+"""
+
+
+RasterParse = namedtuple("RasterParse", "ox oy width height depth image",
+                         module="matplotlib.mathtext")
+RasterParse.__doc__ = r"""
+The namedtuple type returned by ``MathTextParser("agg").parse(...)``.
+
+This tuple contains the global metrics (*width*, *height*, *depth*), and a
+raster *image*.  The offsets *ox*, *oy* are always zero.
+"""
+
+
+class Output:
+    r"""
+    Result of `ship`\ping a box: lists of positioned glyphs and rectangles.
+
+    This class is not exposed to end users, but converted to a `VectorParse` or
+    a `RasterParse` by `.MathTextParser.parse`.
+    """
+
+    def __init__(self, box):
+        self.box = box
+        self.glyphs = []  # (ox, oy, info)
+        self.rects = []  # (x1, y1, x2, y2)
+
+    def to_vector(self):
+        w, h, d = map(
+            np.ceil, [self.box.width, self.box.height, self.box.depth])
+        gs = [(info.font, info.fontsize, info.num, ox, h - oy + info.offset)
+              for ox, oy, info in self.glyphs]
+        rs = [(x1, h - y2, x2 - x1, y2 - y1)
+              for x1, y1, x2, y2 in self.rects]
+        return VectorParse(w, h + d, d, gs, rs)
+
+    def to_raster(self):
+        xmin = min([*[ox + info.metrics.xmin for ox, oy, info in self.glyphs],
+                    *[x1 for x1, y1, x2, y2 in self.rects], 0]) - 1
+        ymin = min([*[oy - info.metrics.ymax for ox, oy, info in self.glyphs],
+                    *[y1 for x1, y1, x2, y2 in self.rects], 0]) - 1
+        xmax = max([*[ox + info.metrics.xmax for ox, oy, info in self.glyphs],
+                    *[x2 for x1, y1, x2, y2 in self.rects], 0]) + 1
+        ymax = max([*[oy - info.metrics.ymin for ox, oy, info in self.glyphs],
+                    *[y2 for x1, y1, x2, y2 in self.rects], 0]) + 1
+        w = xmax - xmin
+        h = ymax - ymin - self.box.depth
+        d = ymax - ymin - self.box.height
+        image = FT2Image(np.ceil(w), np.ceil(h + max(d, 0)))
+
+        # Ideally, we could just use self.glyphs and self.rects here, shifting
+        # their coordinates by (-xmin, -ymin), but this yields slightly
+        # different results due to floating point slop; shipping twice is the
+        # old approach and keeps baseline images backcompat.
+        shifted = ship(self.box, (-xmin, -ymin))
+
+        for ox, oy, info in shifted.glyphs:
+            info.font.draw_glyph_to_bitmap(
+                image, ox, oy - info.metrics.iceberg, info.glyph,
+                antialiased=mpl.rcParams['text.antialiased'])
+        for x1, y1, x2, y2 in shifted.rects:
+            height = max(int(y2 - y1) - 1, 0)
+            if height == 0:
+                center = (y2 + y1) / 2.0
+                y = int(center - (height + 1) / 2.0)
```

Comment:
Usually you fix these :stuck_out_tongue: 
```suggestion
                center = (y2 + y1) / 2
                y = int(center - (height + 1) / 2)
```

Reply:
done